ethers-providers/src/provider.rs

```diff
@@ -1161,6 +1161,278 @@ impl<P: JsonRpcClient> Middleware for Provider<P> {
     }
 }
 
+use crate::PinBoxFut;
+use serde::ser::SerializeTuple;
+use std::{
+    collections::HashMap,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+impl<P: JsonRpcClient> Provider<P> {
+    pub fn call_builder<'a>(&'a self, tx: &'a TypedTransaction) -> Caller<'a, P> {
+        Caller::Build(CallBuilder::new(self, tx))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CallInput<'a> {
+    tx: &'a TypedTransaction,
+    block: Option<BlockId>,
+    spoof: Option<&'a spoof::State>,
+}
+impl<'a> CallInput<'a> {
+    pub fn new(tx: &'a TypedTransaction) -> Self {
+        Self { tx, block: None, spoof: None }
+    }
+}
+
+impl<'a> Serialize for CallInput<'a> {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        let len = 2 + self.spoof.is_some() as usize;
+
+        let mut tup = serializer.serialize_tuple(len)?;
+        tup.serialize_element(self.tx)?;
+
+        let block = self.block.unwrap_or_else(|| BlockNumber::Latest.into());
+        tup.serialize_element(&block)?;
+
+        if let Some(spoof) = self.spoof {
+            tup.serialize_element(spoof)?;
+        }
+        tup.end()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct CallBuilder<'a, P> {
+    provider: &'a Provider<P>,
+    input: CallInput<'a>,
+}
+
+impl<'a, P> CallBuilder<'a, P> {
+    fn new(provider: &'a Provider<P>, tx: &'a TypedTransaction) -> Self {
+        Self { provider, input: CallInput::new(tx) }
+    }
+}
+impl<'a, P: JsonRpcClient> CallBuilder<'a, P> {
+    fn execute(&self) -> impl Future<Output = Result<Bytes, ProviderError>> + 'a {
```

Comment:
this is not async because this is the state of the caller?

Reply:
The short answer is I tried async briefly and upset the borrow checker.

```rust
async fn execute(&self) -> Result<Bytes, ProviderError>;
```

As I understand it, the problem is that this causes the returned future to capture the anonymous lifetime `'self`. `Caller::Wait(PinBoxFut<'a, Bytes>)` needs to hold a future tied to the `'a` lifetime of the reference to the `Provider`. So making `execute()` async and attempting to store this future in `Future::poll` pulls in the `Pin<&mut Self>` lifetime which can't match the `'a` provider lifetime.